File: src/data_utils/load_mnist.py
import numpy as np
from tensorflow.keras import utils


def load_MNIST(
    file_path="data/mnist.txt",
    train_size=60000,
    test_size=10000,
    to_categorical=False,
    normalize=False
):
    NUM_FEATURES = 8 * 8
    NUM_CLASSES = 10
    OG_TRAIN_SIZE = 60000
    OG_TEST_SIZE = 10000

    if test_size is None:
        test_size = OG_TEST_SIZE

    assert train_size >= 0 and train_size <= OG_TRAIN_SIZE
    assert test_size >= 0 and test_size <= OG_TEST_SIZE

    # Read and parse dataset from file.
    with open(file_path) as f:
        lines = f.readlines()
    x = [[int(c) for c in l[:NUM_FEATURES]] for l in lines]
    y = [int(l[NUM_FEATURES]) for l in lines]
    train_x = np.array(x[:train_size])
    train_y = np.array(y[:train_size])
    test_x = np.array(x[OG_TRAIN_SIZE:OG_TRAIN_SIZE+test_size])
    test_y = np.array(y[OG_TRAIN_SIZE:OG_TRAIN_SIZE+test_size])

    # Check shapes.
    assert train_size == 0 or train_x.shape == (train_size, NUM_FEATURES)
    assert train_size == 0 or train_y.shape == (train_size,)
    assert test_size == 0 or test_x.shape == (test_size, NUM_FEATURES)
    assert test_size == 0 or test_y.shape == (test_size,)

    # Turn labels to one-hot encoding to fit neural network output layer.
    if to_categorical:
        train_y = utils.to_categorical(train_y, NUM_CLASSES)
        test_y = utils.to_categorical(test_y, NUM_CLASSES)

    return train_x, train_y, test_x, test_y


if __name__ == "__main__":

    import matplotlib.pyplot as pp

    train_x, train_y, test_x, test_y = load_MNIST(
        file_path="data/mnist.txt",
        train_size=100,
        test_size=100,
        to_categorical=True
    )

    # Show a sample to check everything is fine.
    id = 0
    print("label", train_y[id])
    pp.imshow(train_x[id].reshape((8, 8)), cmap='gray')
    pp.show()

File: src/data_utils/load_mnist.cpp
#include "load_mnist.h"

#include <iostream>
#include <fstream>
#include <string>



void loadMNIST(
    std::vector<std::vector<int64_t>> &trainX,
    std::vector<int64_t> &trainY,
    std::vector<std::vector<int64_t>> &testX,
    std::vector<int64_t> &testY
)
{
    std::string line;
    std::ifstream inputStream;
    std::vector<int64_t> sample;
    size_t id, i;

    inputStream.open("data/mnist.txt");
    if (!inputStream.is_open())
        std::cerr << "Error opening file" << std::endl;

    id = 0;
    while (std::getline(inputStream, line)) {
        for (i = 0; i < line.length() - 1; i++)
            sample.push_back(line[i] - '0');
        if (id < 60000) {
            trainX.push_back(sample);
            trainY.push_back(line[i] - '0');
        } else {
            testX.push_back(sample);
            testY.push_back(line[i] - '0');
        }
        sample.clear();
        id++;
    }

    inputStream.close();
}


void loadMNIST(
    std::vector<std::vector<double>> &trainX,
    std::vector<std::vector<double>> &trainY,
    std::vector<std::vector<double>> &testX,
    std::vector<std::vector<double>> &testY,
    size_t trainSize,
    size_t testSize,
    size_t trainOffset,
    size_t testOffset
)
{
    std::vector<std::vector<int64_t>> trainXint, testXint;
    std::vector<int64_t> trainYint, testYint;
    loadMNIST(trainXint, trainYint, testXint, testYint);

    for (size_t i = trainOffset; i < trainOffset + trainSize; i++)
    {
        std::vector<int64_t> xint = trainXint[i];
        std::vector<double> x(xint.begin(), xint.end());
        trainX.push_back(x);

        std::vector<double> y(10);
        y[trainYint[i]] = 1.0;
        trainY.push_back(y);
    }

    for (size_t i = testOffset; i < testOffset + testSize; i++)
    {
        std::vector<int64_t> xint = testXint[i];
        std::vector<double> x(xint.begin(), xint.end());
        testX.push_back(x);

        std::vector<double> y(10);
        y[testYint[i]] = 1.0;
        testY.push_back(y);
    }
}


// int main(int argc, char* argv[]) {
    
//     std::vector<std::vector<int64_t>> trainX, testX;
//     std::vector<int64_t> trainY, testY;

//     loadMNIST(trainX, trainY, testX, testY);

//     for (size_t i = 0; i < trainX.size(); i++) {
//         std::cout << i << " ";
//         for (auto value : trainX[i])
//             std::cout << value;
//         std::cout << trainY[i] << std::endl;
//     }

//     for (size_t i = 0; i < testX.size(); i++) {
//         std::cout << i << " ";
//         for (auto value : testX[i])
//             std::cout << value;
//         std::cout << testY[i] << std::endl;
//     }

// }
